### ISSUE TITLE
From paths to "modules"

### ISSUE THREAD

**Issue by Timoses** [kubernetes-sigs/kustomize#1100]
I have began using kustomize for just over a week now and feel like the many many "../../../" in the kustomization.yaml files are too distracting. It feels very hard to follow where files are located and what project they are referencing. Take for example:

in `projA/helperX/base/kustomization.yaml`:
```
bases:
- ../../crds
- ../base
```

Would it not be nice to have something like a "modular" notation like the following:

in `projA/helperX/base/kustomization.yaml`:
```
bases:
- projA.app1.crds
- projA.helperX.base
```

I suppose this introduces the question of introducing the notion of a root directory.

Large kustomize projects could become unwieldy quickly I'm afraid. With an approach of naming the involved "modules" it could become easier to track what is happening.

**Comment by bittner** [comment]
I agree.

> I suppose this introduces the question of introducing the notion of a root directory.

True.

Somewhat related: Some kustomize commands (e.g. `kustomize edit add configmap ...`) seem to assume you're running it from a folder that contains a *kustomization.yaml* file:
```
Error: Missing kustomization file 'kustomization.yaml'.
```

It should be possible to refer to (a folder containing) a kustomization and kustomize would deduce from the setup what the related base directory is.

**Comment by Timoses** [comment]
> Somewhat related: Some kustomize commands (e.g. kustomize edit add configmap ...) seem to assume you're running it from a folder that contains a kustomization.yaml file:

I rarely use `kustomize` commands to edit `kustomization` files. One could add a `root` to a kustomization like:
```
root: ../../../
```

However, this would have to be added to all kustomization files in the project which is
1. ugly
2. making mistakes becomes easy (drop one `../` in one kustomization file)

**Comment by Liujingfang1** [comment]
This feature would help #1119.

**Comment by bittner** [comment]
> I rarely use `kustomize` commands to edit `kustomization` files. One could add a `root` to a kustomization

Note that the `root` is implicit when you run `kustomize build the/path/to/your/root` (i.e. specify a target path explicitly). No need to add a new language item.

**Comment by Timoses** [comment]
After some time with `kustomize` I still think something like a named modular approach would be great.

Although I have come to achieve a good structure of the k8s config repository using `kustomize`, it would sometimes be practical to be able to reuse some resources over a wide range of directories or even other repositories.

For example, if I am using many `Ingress` resources in my k8s config directory, it could be beneficial to simply use something like:
```
modules:
- ingress/type1.yaml

patchesStrategicMerge:
- overlays/ingress.yaml
```

`ingress/type1.yaml` would be reusable in many different resources. I'd only ever need to create an overlay specifying the customizable fields (hosts, tls, rules).

The question only is: Where does it come from?
Perhaps there could be a `module_dir` directive which points to a directory containing "modules" (resources that are meant to be used by many applications / kustomize directories).
So perhaps:
```
module_dir: ../../../../     # could also be a directory in a git repository: https://github.com/MyCompany/utils/tree/master/k8s_modules

modules:
- ingress/type1.yaml
```
